packages/cosmic-proto/src/address-hooks.js

```diff
@@ -0,0 +1,231 @@
+ *
+ * `magic` is a 3-byte prefix that identifies a hooked address and its version
+ * nibble, whose value is 4 bits (between 0 and 0xf (15)).  Currently, the only
+ * supported version is 0.
+ *
+ *      0x78, 0xf1, (0x70 | ADDRESS_HOOK_VERSION),
+ *
+ * This magic prefix encodes as `0rch`, regardless of the version or HRP (e.g.
+ * `agoric10rch<rest of payload as bech32><bech32 checksum>`).
+ *
+ * @param {string} baseAddress
+ * @param {ArrayLike<number>} hookData
+ * @param {number} [charLimit]
+ * @returns {string}
+ */
+export const joinHookedAddress = (
+  baseAddress,
+  hookData,
+  charLimit = DEFAULT_HOOKED_ADDRESS_CHAR_LIMIT,
+) => {
+  const { prefix, bytes } = decodeBech32(baseAddress, charLimit);
+
+  const baseAddressLength = bytes.length;
+  if (baseAddressLength > MAX_BASE_ADDRESS_LENGTH) {
+    throw RangeError(
+      `Base address length 0x${baseAddressLength.toString(16)} exceeds maximum 0x${MAX_BASE_ADDRESS_LENGTH.toString(16)}`,
+    );
+  }
+  const b = baseAddressLength;
+
+  const magicLength = ADDRESS_HOOK_MAGIC.length;
+  const hookBuf = new Uint8Array(
+    magicLength + bytes.length + hookData.length + BASE_ADDRESS_LENGTH_BYTES,
+  );
+  hookBuf.set(ADDRESS_HOOK_MAGIC);
+  hookBuf.set(bytes, magicLength);
+  hookBuf.set(hookData, magicLength + b);
+
+  // Append the big-endian address length.
+  let len = b;
+  for (let i = 0; i < BASE_ADDRESS_LENGTH_BYTES; i += 1) {
+    hookBuf[hookBuf.length - 1 - i] = len & 0xff;
+    len >>>= 8;
+  }
+
+  return encodeBech32(prefix, hookBuf, charLimit);
+};
+
+/**
+ * @param {string} baseAddress
+ * @param {HookQuery} query
+ * @param {number} [charLimit]
+ */
+export const encodeAddressHook = (baseAddress, query, charLimit) => {
+  const queryStr = queryString.stringify(query);
+
+  const te = new TextEncoder();
+  const hookData = te.encode(`?${queryStr}`);
+  return joinHookedAddress(baseAddress, hookData, charLimit);
+};
+
+/**
+ * @param {string} addressHook
+ * @param {number} [charLimit]
+ * @returns {{ baseAddress: string; query: HookQuery }}
+ */
+export const decodeAddressHook = (addressHook, charLimit) => {
+  const { baseAddress, hookData } = splitHookedAddress(addressHook, charLimit);
+  const hookStr = new TextDecoder().decode(hookData);
+  if (hookStr && !hookStr.startsWith('?')) {
+    throw Error(`Hook data does not start with '?': ${hookStr}`);
+  }
+
+  /** @type {HookQuery} */
+  const query = queryString.parse(hookStr);
+  return { baseAddress, query };
+};
+
+/**
+ * @param {string} specimen
+ * @param {number} [charLimit]
+ * @returns {string | { baseAddress: string; hookData: Uint8Array }}
+ */
+export const splitHookedAddressUnsafe = (
+  specimen,
+  charLimit = DEFAULT_HOOKED_ADDRESS_CHAR_LIMIT,
+) => {
+  const { prefix, bytes } = decodeBech32(specimen, charLimit);
+
+  const magicLength = ADDRESS_HOOK_MAGIC.length;
+  for (let i = 0; i < magicLength; i += 1) {
+    if (bytes[i] !== ADDRESS_HOOK_MAGIC[i]) {
```

Comment:
it still blows me away that the JS stdlib doesn't include array comparison operators.